ddtrace/ext/ci_visibility/_ci_visibility_base.py

```diff
@@ -0,0 +1,98 @@
+import abc
+import dataclasses
+from enum import Enum
+from typing import Any
+from typing import Dict
+from typing import Generic
+from typing import List
+from typing import Optional
+from typing import TypeVar
+from typing import Union
+
+from ddtrace.internal.logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class _CIVisibilityRootItemIdBase:
+    """This class exists for the ABC class below"""
+
+    name: str
+
+    def get_parent_id(self) -> "_CIVisibilityRootItemIdBase":
+        return self
+
+    def get_session_id(self) -> "_CIVisibilityRootItemIdBase":
+        return self
+
+
+RT = TypeVar("RT", bound="_CIVisibilityRootItemIdBase")
+
+
+@dataclasses.dataclass(frozen=True)
+class _CIVisibilityIdBase(abc.ABC):
+    @abc.abstractmethod
+    def get_parent_id(self) -> Union["_CIVisibilityIdBase", _CIVisibilityRootItemIdBase]:
+        raise NotImplementedError("This method must be implemented by the subclass")
+
+    def get_session_id(self) -> _CIVisibilityRootItemIdBase:
+        return self.get_parent_id().get_session_id()
+
+
+PT = TypeVar("PT", bound=Union[_CIVisibilityIdBase, _CIVisibilityRootItemIdBase])
+
+
+@dataclasses.dataclass(frozen=True)
+class _CIVisibilityChildItemIdBase(_CIVisibilityIdBase, Generic[PT]):
+    parent_id: PT
+    name: str
+
+    def get_parent_id(self) -> PT:
+        return self.parent_id
+
+
+CIItemId = TypeVar("CIItemId", bound=Union[_CIVisibilityChildItemIdBase, _CIVisibilityRootItemIdBase])
+
+
+class _CIVisibilityAPIBase(abc.ABC):
+    def __init__(self):
+        raise NotImplementedError("This class is not meant to be instantiated")
+
+    @staticmethod
+    @abc.abstractmethod
+    def discover(item_id: _CIVisibilityRootItemIdBase, *args, **kwargs):
+        pass
+
+    @staticmethod
+    @abc.abstractmethod
+    def start(item_id: _CIVisibilityRootItemIdBase, *args, **kwargs):
+        pass
+
+    @staticmethod
+    @abc.abstractmethod
+    def finish(
+        item_id: _CIVisibilityRootItemIdBase,
+        override_status: Optional[Enum],
+        force_finish_children: bool = False,
+        *args,
+        **kwargs,
+    ):
+        pass
+
+    @staticmethod
+    def set_tag(item_id: _CIVisibilityRootItemIdBase, tag_name: str, tag_value: Any, recurse: bool = False):
+        log.debug("Setting tag for item %s: %s=%s", item_id, tag_name, tag_value)
+
+    @staticmethod
+    def set_tags(item_id: _CIVisibilityRootItemIdBase, tags: Dict[str, Any], recurse: bool = False):
+        log.debug("Setting tags for item %s: %s", item_id, tags)
+
+    @staticmethod
+    def delete_tag(item_id: _CIVisibilityRootItemIdBase, tag_name: str, recurse: bool = False):
+        log.debug("Deleting tag for item %s: %s", item_id, tag_name)
+
+    @staticmethod
+    def delete_tags(item_id: _CIVisibilityRootItemIdBase, tag_names: List[str], recurse: bool = False):
+        log.debug("Deleting tags for item %s: %s", item_id, tag_names)
```

Comment:
This logging might be confusing as a default implementation that doesn't actually do the thing it's logging about.

Reply:
Yeah this falls in the "not implemented yet" set of things in the API.